components/script/dom/servoparser/async_html.rs

```diff
@@ -0,0 +1,553 @@
+        debug!("Parse error: {}", msg);
+    }
+
+    fn set_quirks_mode(&mut self, mode: QuirksMode) {
+        let mode = match mode {
+            QuirksMode::Quirks => ServoQuirksMode::Quirks,
+            QuirksMode::LimitedQuirks => ServoQuirksMode::LimitedQuirks,
+            QuirksMode::NoQuirks => ServoQuirksMode::NoQuirks,
+        };
+        self.document.set_quirks_mode(mode);
+    }
+
+    fn append(&mut self, parent: &Self::Handle, child: NodeOrText<Self::Handle>) {
+        let new_node = match child {
+            NodeOrText::AppendNode(node) => ParserNodeOrText::Node(node.id),
+            NodeOrText::AppendText(text) => ParserNodeOrText::Text(text.into()),
+        };
+        self.enqueue(ParseOperation::Insert(parent.id, None, new_node));
+    }
+
+    fn append_doctype_to_document(&mut self, name: StrTendril, public_id: StrTendril,
+                                  system_id: StrTendril) {
+        self.enqueue(ParseOperation::AppendDoctypeToDocument(name.into(), public_id.into(), system_id.into()));
+    }
+
+    fn add_attrs_if_missing(&mut self, target: &Self::Handle, attrs: Vec<Attribute>) {
+        self.enqueue(ParseOperation::AddAttrsIfMissing(target.id, attrs));
+    }
+
+    fn remove_from_parent(&mut self, target: &Self::Handle) {
+        let mut parse_node_data = self.parse_node_data.borrow_mut();
+        let data = parse_node_data.get_mut(&target.id).expect("Element does not exist!");
+        data.parent = None;
+        self.enqueue(ParseOperation::RemoveFromParent(target.id));
+    }
+
+    fn mark_script_already_started(&mut self, node: &Self::Handle) {
+        self.enqueue(ParseOperation::MarkScriptAlreadyStarted(node.id));
+    }
+
+    // TODO: Handle this
+    fn complete_script(&mut self, node: &Self::Handle) -> NextParserState {
+        // if let Some(script) = node.downcast() {
+        //     self.script.set(Some(script));
+        //     NextParserState::Suspend
+        // } else {
+        //     NextParserState::Continue
+        // }
+        NextParserState::Suspend
+    }
+
+    fn reparent_children(&mut self, node: &Self::Handle, new_parent: &Self::Handle) {
+        for data in self.parse_node_data.borrow_mut().values_mut() {
+            if data.parent == Some(node.id) {
+                data.parent = Some(new_parent.id);
+            }
+        }
+        self.enqueue(ParseOperation::ReparentChildren(node.id, new_parent.id));
+    }
+
+    /// https://html.spec.whatwg.org/multipage/#html-integration-point
+    /// Specifically, the <annotation-xml> cases.
+    fn is_mathml_annotation_xml_integration_point(&self, handle: &Self::Handle) -> bool {
+        let parse_node_data = self.parse_node_data.borrow();
+        let node_data = parse_node_data.get(&handle.id).expect("Element does not exist!");
+        node_data.is_integration_point
+    }
+
+    fn set_current_line(&mut self, line_number: u64) {
+        self.current_line = line_number;
```

Comment:
I misread the changes. Currently this works fine, but when we stop calling process_operation immediately after creating parse nodes it will no longer work.